Add blog post for KMSv2 beta (#40194)
kubernetes/website: content/en/blog/_posts/2023-04-11-kmsv2-beta.md

---
layout: blog
title: "Kubernetes 1.27: KMS V2 Moves to Beta"
date: 2023-04-11
slug: kms-v2-moves-to-beta
---

**Authors:** TBD

With Kubernetes 1.27, we are moving Key Management Service (KMS) v2 API to beta.

## What is KMS?

> One of the first things to consider when securing a Kubernetes cluster is encrypting etcd data at rest. KMS provides an interface for a provider to utilize a key stored in an external key service to perform this encryption.

Reviewer: Please **manually wrap** the long lines like this for ease of review and for ease of change tracking, especially for downstream localization teams. I.e.,

```suggestion

One of the first things to consider when securing a Kubernetes cluster is encrypting etcd data at rest.
KMS provides an interface for a provider to utilize a key stored in an external key service
to perform this encryption.
```

> KMS v1 has been a feature of Kubernetes since version 1.10, and is currently in beta as of version v1.12. KMS v2 was introduced as alpha in v1.25.

Reviewer: ```suggestion
KMS v1 has been a feature of Kubernetes since version 1.10,
and is currently in beta since version v1.12.
KMS v2 was introduced as alpha in v1.25.
```

> {{< note >}} The design in [the previous blog post](https://kubernetes.io/blog/2022/09/09/kms-v2-improvements/) is not compatible with the design in this blog post. The KMS v2 API and implementation changed in incompatible ways in-between the alpha release in v1.25 and the beta release in v1.27. Attempting to upgrade from old versions with the alpha feature enabled will result in data loss. {{< /note >}}

Reviewer: ```suggestion
{{< note >}}
The design in [the previous blog post](https://kubernetes.io/blog/2022/09/09/kms-v2-improvements/)
is not compatible with the design in this blog post. The KMS v2 API and implementation changed
in incompatible ways in-between the alpha release in v1.25 and the beta release in v1.27.
Attempting to upgrade from old versions with the alpha feature enabled will result in data loss.
{{< /note >}}
```

Reviewer 2: Ideally, avoid using callouts (note, warning, etc) in blog articles. This is for a couple of reasons, one is maintainability (it's more to redo if we ever change how callouts work) and the other is that blogs aren't docs.

Using plain Markdown, perhaps with an emoji - eg :warning: - is how I'd do it.

---

Site relative links are better still:
```markdown
The design in the [previous blog post](/blog/2022/09/09/kms-v2-improvements/)
is not compatible with the design covered in this blog post.  …
```

Author: thanks for the feedback, @sftim! the caution called out here is something we **really** want to ensure is read before using the feature. We also have this in the docs but wanted to echo this in the blog too. Having said that, we're trying to use the emoji option but the current rendered docs with the callouts (https://deploy-preview-40194--kubernetes-io-main-staging.netlify.app/blog/2023/04/11/kms-v2-moves-to-beta/) look much cleaner. Is there any chance we could use it for the blog? The note we might be able to remove the callout and just italicize the text.

Reviewer 2: Try a Docsy [alert](https://www.docsy.dev/docs/adding-content/shortcodes/#alert). There are a few examples of these around the site already.

Author: the Docsy alert looks good! thanks for the suggestion

## What’s new in `v2beta1`?

> The KMS encryption provider uses an envelope encryption scheme to encrypt data in etcd. The data is encrypted using a data encryption key (DEK). The DEKs are encrypted with a key encryption key (KEK) that is stored and managed in a remote KMS. With KMS v1, a new DEK is generated for each encryption. With KMS v2, a new DEK is generated on server startup and when the KMS plugin informs the API server that a KEK rotation has occurred.

Reviewer 3: ```suggestion
The KMS encryption provider uses an envelope encryption scheme to encrypt data in etcd. The data is encrypted using a data encryption key (DEK). The DEKs are encrypted with a key encryption key (KEK) that is stored and managed in a remote KMS. With KMS v1, a new DEK is generated for each encryption. With KMS v2, a new DEK is only generated on server startup and when the KMS plugin informs the API server that a KEK rotation has occurred.
```

{{% alert title="Caution" color="warning" %}}
If you are running virtual machine (VM) based nodes that leverage VM state store with this feature, you must not use KMS v2.

With KMS v2, the API server uses AES-GCM with a 12 byte nonce (8 byte atomic counter and 4 bytes random data) for encryption. The following issues could occur if the VM is saved and restored:

1. The counter value may be lost or corrupted if the VM is saved in an inconsistent state or restored improperly. This can lead to a situation where the same counter value is used twice, resulting in the same nonce being used for two different messages.
2. If the VM is restored to a previous state, the counter value may be set back to its previous value, resulting in the same nonce being used again.

> Although both of these cases are partially mitigated by the 4 byte random nonce, this can compromise the security of the encryption.

Reviewer 4: ```suggestion

Although both of these cases are partially mitigated by the 4 byte random nonce, this can compromise the security of the encryption.
```

        kube_api_server->>kube_api_server: generate DEK
        kube_api_server->>kms_plugin: encrypt request
        kms_plugin->>external_kms: encrypt DEK with remote KEK
        external_kms->>kms_plugin: encrypted DEK
        kms_plugin->>kube_api_server: return encrypt response <br/> {"ciphertext": "<encrypted DEK>", key_id: "<remote KEK ID>", "annotations": {}}
```
-->

{{< figure src="/images/blog/2023-04-11-kubernetes-1.27-kmsv2-beta/kubernetes-1.27-generate-dek.svg" alt="Sequence diagram for KMSv2 beta Generate DEK" class="diagram-large" >}}

### Performance Improvements
With KMS v2, we have made significant improvements to the performance of the KMS encryption provider.
In case of KMS v1, a new DEK is generated for every encryption. This means that for every write request, the API server makes a call to the KMS plugin to encrypt the DEK using the remote KEK. The API server also has to cache the DEKs to avoid making a call to the KMS plugin for every read request. When the API server restarts, it has to populate the cache by making a call to the KMS plugin for every DEK in the etcd store based on the cache size. This is a significant overhead for the API server.

> With KMS v2, the API server generates a DEK at startup and caches it. The API server also makes a call to the KMS plugin to encrypt the DEK using the remote KEK. This is a one-time call at startup. The API server then uses the cached DEK to encrypt the resources. This reduces the number of calls to the KMS plugin and improves the overall latency of the API server requests.

Reviewer 4: ```suggestion
With KMS v2, the API server generates a DEK at startup and caches it. The API server also makes a call to the KMS plugin to encrypt the DEK using the remote KEK. This is a one-time call at startup and on KEK rotation. The API server then uses the cached DEK to encrypt the resources. This reduces the number of calls to the KMS plugin and improves the overall latency of the API server requests.
```

> We conducted a test creating 12k secrets and measuring the time taken for the API server to encrypt the resources. The metric used was [`apiserver_storage_transformation_duration_seconds`](https://kubernetes.io/docs/reference/instrumentation/metrics/).

Reviewer 4: ```suggestion
We conducted a test that created 12k secrets and measured the time taken for the API server to encrypt the resources. The metric used was [`apiserver_storage_transformation_duration_seconds`](https://kubernetes.io/docs/reference/instrumentation/metrics/).
```

---
layout: blog
title: "Kubernetes 1.27: KMS V2 Moves to Beta"

> date: 2023-04-11

Reviewer 5: ```suggestion
date: 2023-05-01
```

Reviewer 2: Please also change the path `content/en/blog/_posts/2023-04-11-kmsv2-beta.md` to match the new date.

Author: done!

slug: kms-v2-moves-to-beta
---

**Authors:** Anish Ramasekar, Mo Khan, and Rita Zhang (Microsoft)

> With Kubernetes 1.27, we are moving Key Management Service (KMS) v2 API to beta.

Reviewer 2: Is “we” Microsoft, or Kubernetes? Consider making it clear.

Author: Updated it to make it clear. I used the similar pattern that was used in few other blogs. LMK if this is ok!

#### Status Request

<!-- source
```mermaid
%%{init:{"theme":"neutral", "sequence": {"mirrorActors":true},
    "themeVariables": {
        "actorBkg":"royalblue",
        "actorTextColor":"white"
}}}%%

sequenceDiagram
    participant kube_api_server
    participant kms_plugin
    participant external_kms
    alt Generate DEK at startup
        Note over kube_api_server,external_kms: Refer to Generate Data Encryption Key (DEK) diagram for details
    end
    loop every minute (or every 10s if error or unhealthy)
        kube_api_server->>kms_plugin: status request
        kms_plugin->>external_kms: validate remote KEK
        external_kms->>kms_plugin: KEK status
        kms_plugin->>kube_api_server: return status response <br/> {"healthz": "ok", key_id: "<remote KEK ID>", "version": "v2beta1"}
        alt KEK rotation detected (key_id changed), rotate DEK
            Note over kube_api_server,external_kms: Refer to Generate Data Encryption Key (DEK) diagram for details
        end
    end
```
-->

{{< figure src="/images/blog/2023-04-11-kubernetes-1.27-kmsv2-beta/kubernetes-1.27-status.svg" alt="Sequence diagram for KMSv2 beta Status" class="diagram-large" >}}

> #### Generate Data Encryption Key (DEK)

Reviewer 3: There's nothing in this diagram that shows the reuse of DEK and a new DEK is only generated on server startup and a KEK rotation. I think this is the big value add for v2beta1.

Author: `Generate Data Encryption Key (DEK)` is to only show what the flow of events is when a DEK is generated (during startup or rotation of KEK).  This was used in multiple places, so moved it a single diagram and this is referenced in the ones above. 

The [sequence diagram for Encrypt](https://deploy-preview-40194--kubernetes-io-main-staging.netlify.app/blog/2023/04/11/kms-v2-moves-to-beta/#encrypt-request) shows the DEK is only generated at startup, the [diagram for status](https://deploy-preview-40194--kubernetes-io-main-staging.netlify.app/blog/2023/04/11/kms-v2-moves-to-beta/#status-request) shows in addition to startup, the DEK being generated when the KEK is rotated.

Reviewer 3: Ah ok. my eyes went straight to the gray box and didnt see the `[]` text above it.

We conducted a test that created 12k secrets and measured the time taken for the API server to encrypt the resources. The metric used was [`apiserver_storage_transformation_duration_seconds`](https://kubernetes.io/docs/reference/instrumentation/metrics/).
For KMS v1, the test was run on a managed Kubernetes v1.25 cluster with 2 nodes. There was no additional load on the cluster during the test.
For KMS v2, the test was run in the Kubernetes CI environment with the following [cluster configuration](https://github.com/kubernetes/kubernetes/blob/release-1.27/test/e2e/testing-manifests/auth/encrypt/kind.yaml).

| KMS Provider | Time taken by 95 percentile |
| ------------ | --------------------------- |
| KMS v1       | 160ms                       |
| KMS v2       | 80μs                        |

The results show that the KMS v2 encryption provider is three orders of magnitude faster than the KMS v1 encryption provider.

## What's next?
For Kubernetes v1.28, we expect the feature to stay in beta. In the coming releases we want to investigate:
- Cryptographic changes to remove the limitation on VM state store.

> - Kubernetes rest API changes to enable a more robust story around key rotation.

Reviewer 6: ```suggestion
- Kubernetes REST API changes to enable a more robust story around key rotation.
```
